prevented non-numbers from being entered in date fields

aliceriot opened the pull request:
#### What are the relevant tickets?

closes #585 and closes #586 

#### What's this PR do?

This changes the text fields in `boundDateField` so that they only accept characters that can comprise valid inputs. So, no non-numerical characters, no characters beyond the number required for the field in question, etc.

#### Where should the reviewer start?

Review the changes to `profile_edit.js` and `validation.js`. 

#### How should this be manually tested?

Make sure you can edit and save valid dates for all date fields we have.

Also, check for the following behavior, which this PR implements:

month field:

- should accept only 2 characters
- should not accept non-numerical characters
- should only allow input of 1-12
- input of numbers larger than 12 should correct to 12

day field:

- should accept only 2 characters
- should not accept non-numerical characters
- should only allow input of 1-31
- input of numbers larger than 31 should correct to 31

year field:

- should accept 4 characters max
- should not accept non-numerical characters
- should accept numbers below 1800 as input, but they should be marked invalid on save
- should accept any number `x | 1800 <= x <= 2100` as valid input
- input of numbers larger than 2100 should correct to 2100

#### Any background context you want to provide?

#### Screenshots (if appropriate)

#### What GIF best describes this PR or how it makes you feel?

noisecapella commented:
Can you change the PR title to the past tense?

noisecapella commented:
If the month field is empty I can enter in any number of non-numeric characters. It only seems to do the check if there is a number in the field already

aliceriot commented:
@noisecapella I fixed the bug with empty fields and added left-padding on month and day values. I don't think there's a good way to add left-padding on values as the user is typing, so it will only pad when there's no `edit` object on the profile for the field in question (meaning the field will be left padded when you open the edit form for, say, work history, but editing the field will erase the left padding and allow you to edit the number freely.

noisecapella commented:
I'm seeing an edge case during testing: 
 - Go to a field where the date was already populated. Let's say the year is '1985'.
 - Move the cursor to after the '9' and hit backspace. The cursor will be moved to the end of the string. The string will now be '185' which is as expected

noisecapella commented:
It looks like that edge case is due to `1985` being passed to the component as a number but the edited values being strings

noisecapella commented:
Also can you add a `closes #xxx` for the other issue that's also fixed?

aliceriot commented:
oh huh, I edited the pr comment yesterday to add that but it didn't save, strange.

aliceriot commented:
@noisecapella it looks like the bug with the cursor jumping to the end is present on master, and on all of our text fields (not just date fields).

So big :( there, do you think we should try to fix it here or open a new issue?

noisecapella commented:
Ah, let's just open a new issue

aliceriot commented:
kk will do

noisecapella commented on static/js/util/validation.js:
```diff
@@ -246,44 +246,71 @@ export function validateProfileComplete(profile: Profile): ProfileComplete {
   return date;
 }
 
+export function sanitizeDate(input: string|number, length: number): string {
```
Can you add a JSDoc comment here?

noisecapella commented on static/js/util/validation_test.js:
```diff
@@ -320,29 +321,92 @@ describe('Profile validation functions', () => {
+
+    describe('numerical input', () => {
+      it('should return a string', () => {
+        assert.deepEqual(sanitizeDate(3, 1), '3');
```
Feel free to leave it but we don't need `deepEqual` for strings

noisecapella commented:
:+1: after adding the JSDoc
